src/miner.cpp

```diff
@@ -74,321 +74,406 @@ int64_t UpdateTime(CBlockHeader* pblock, const Consensus::Params& consensusParam
-    
-    uint64_t nBlockTx = 0;
-    unsigned int nBlockSigOps = 100;  // Reserve 100 sigops for miners to use in their coinbase transaction
-    int lastFewTxs = 0;
-    CAmount nFees = 0;
 
+    uint64_t nTxSigOps = iter->GetSigOpCount();
+    if (nBlockSize + nTxSize <= BLOCKSTREAM_CORE_MAX_BLOCK_SIZE) // Enforce the "old" sigops for <= 1MB blocks
     {
-        LOCK2(cs_main, mempool.cs);
-        CBlockIndex* pindexPrev = chainActive.Tip();
-        const int nHeight = pindexPrev->nHeight + 1;
-        // Fill in header
-        pblock->hashPrevBlock  = pindexPrev->GetBlockHash();
-        pblock->nBits          = GetNextWorkRequired(pindexPrev, pblock, chainparams.GetConsensus());        
-        pblock->nTime          = GetAdjustedTime();
-        const int64_t nMedianTimePast = pindexPrev->GetMedianTimePast();
-
-        pblock->nVersion = UnlimitedComputeBlockVersion(pindexPrev, chainparams.GetConsensus(),pblock->nTime);
-        // -regtest only: allow overriding block.nVersion with
-        // -blockversion=N to test forking scenarios
-        if (chainparams.MineBlocksOnDemand())
-            pblock->nVersion = GetArg("-blockversion", pblock->nVersion);
-
-        int64_t nLockTimeCutoff = (STANDARD_LOCKTIME_VERIFY_FLAGS & LOCKTIME_MEDIAN_TIME_PAST)
-                                ? nMedianTimePast
-                                : pblock->GetBlockTime();
-
-        bool fPriorityBlock = nBlockPrioritySize > 0;
-        if (fPriorityBlock) {
-            vecPriority.reserve(mempool.mapTx.size());
-            for (CTxMemPool::indexed_transaction_set::iterator mi = mempool.mapTx.begin();
-                 mi != mempool.mapTx.end(); ++mi)
-            {
-                double dPriority = mi->GetPriority(nHeight);
-                CAmount dummy;
-                mempool.ApplyDeltas(mi->GetTx().GetHash(), dPriority, dummy);
-                vecPriority.push_back(TxCoinAgePriority(dPriority, mi));
-            }
-            std::make_heap(vecPriority.begin(), vecPriority.end(), pricomparer);
+        // BU: be conservative about what is generated
+        if (nBlockSigOps + nTxSigOps >= BLOCKSTREAM_CORE_MAX_BLOCK_SIGOPS)
+        {
+            // BU: so a block that is near the sigops limit might be shorter than it could be if
+            // the high sigops tx was backed out and other tx added.
+            if (nBlockSigOps > BLOCKSTREAM_CORE_MAX_BLOCK_SIGOPS - 2)
+                blockFinished = true;
+            return false;
+        }
+    }
+    else
+    {
+        uint64_t blockMbSize = 1 + ((nBlockSize + nTxSize - 1) / 1000000);
+        if (nBlockSigOps + nTxSigOps > blockMiningSigopsPerMb.value * blockMbSize)
+        {
+            if (nBlockSigOps > blockMiningSigopsPerMb.value * blockMbSize - 2)
```

Comment:
this is preserving behavior in the original code.  You can see the same thing on line 286, and you'll probably find it in classic if you look.  And yes it does try to find a 1 sig-op tx.  Why not?  worst case you just go through the whole mempool